pkg/ucp/store/inmemory/client.go

```diff
@@ -0,0 +1,289 @@
+/*
+Copyright 2023 The Radius Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inmemory
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/radius-project/radius/pkg/ucp/resources"
+	"github.com/radius-project/radius/pkg/ucp/store"
+	"github.com/radius-project/radius/pkg/ucp/store/storeutil"
+	"github.com/radius-project/radius/pkg/ucp/util/etag"
+	"golang.org/x/exp/maps"
+)
+
+var _ store.StorageClient = (*Client)(nil)
+
+// Client is an in-memory implementation of store.StorageClient.
+type Client struct {
+	// mutex is used to synchronize access to the resources map.
+	mutex sync.Mutex
+
+	// resources is a map of resource IDs to their corresponding entries.
+	//
+	// The Get/Save/Delete methods will use the resource ID directly since they only access
+	// a single entry at a time.
+	//
+	// The Query method will iterate over all entries in the map to find the matching ones.
+	resources map[string]entry
+}
+
+// entry stores the commonly-used fields (extracted from the resource ID) for comparison in queries.
+// This is provided for ease of debugging.
+//
+// We use the existing normalization logic to simplify comparisons:
+//
+// - Convert to lowercase
+// - Add leading/trailing slashes.
+//
+// Here's an example:
+//
+//	resource ID: "/planes/radius/local/resourceGroups/my-rg/providers/Applications.Test/testType1/my-resource/testType2/my-child-resource"
+//
+// The entry would be:
+//
+//	rootScope: "/planes/radius/local/resourcegroups/my-rg/"
+//	resourceType: "/applications.test/testtype1/testtype2/"
```

Comment:
```suggestion
//	resourceType: "/applications.test/testtype1/my-resource/"
```

Reply:
This is, I guess, about the child resource types, right?

- **Applications.Test** is the provider.
- **testType1** is the resource type under that provider.
- **my-resource** is the name of the resource of the type mentioned above.
- **testType2** is the child resource type under the **testType1** resource type.
- **my-child-resource** is the name of the resource of type **testType2**.

Reply:
The resource type is the resource type. It doesn't contain the resource name. 

This is not meant to be documentation for what a resource id **is** this is documentation about how the in-memory data store works.